### DISCUSSION TITLE
softmax in symbol api

### DISCUSSION THREAD

**Discussion by aravindhv10** [apache/incubator-mxnet#13087] [Q&A]
Hello,

     I want to have softmax activation with linear regression output, I was able to get this in gluon in the following way:

class CenteredLayer(mx.gluon.nn.HybridSequential):
    def __init__(self, **kwargs):
        super(CenteredLayer, self).__init__(**kwargs)

    def forward(self, x):
        return x.softmax()

net = gluon.nn.HybridSequential()
with net.name_scope():
    net.add ( gluon.nn.Dense ( sizes[1] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[2] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[3] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[4] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[5] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[4] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[3] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[2] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[1] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[0] , activation="relu" ) )
    net.add ( gluon.nn.Dense ( sizes[0]                     ) )
    net.add ( CenteredLayer()                                 )
softmax_cross_entropy = gluon.loss.L2Loss()

But now I want to get this in the symbol c++ api, I donot want the cross entropy softmax output but only softmax activation with linear regression L2 loss. I was unable to find the softmax operator, is there any documentation for this ?

The c snippet is :

auto input_data   = Symbol::Variable ( "data"       ) ;
                auto target_label = Symbol::Variable ( "data_label" ) ;

                auto L1 = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[1])
                            .SetParam("no_bias",false)
                            .SetInput("data",input_data)
                            .CreateSymbol("L1");
                auto A1 = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L1)
                            .CreateSymbol("A1");

                auto L2 = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[2])
                            .SetParam("no_bias",false)
                            .SetInput("data",A1)
                            .CreateSymbol("L2");
                auto A2 = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L2)
                            .CreateSymbol("A2");

                auto L3 = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[3])
                            .SetParam("no_bias",false)
                            .SetInput("data",A2)
                            .CreateSymbol("L3");
                auto A3 = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L3)
                            .CreateSymbol("A3");

                auto L4 = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[4])
                            .SetParam("no_bias",false)
                            .SetInput("data",A3)
                            .CreateSymbol("L4");
                auto A4 = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L4)
                            .CreateSymbol("A4");

                auto L5 = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[5])
                            .SetParam("no_bias",false)
                            .SetInput("data",A4)
                            .CreateSymbol("L5");
                auto A5 = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L5)
                            .CreateSymbol("A5");

                auto L5Bar = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[4])
                            .SetParam("no_bias",false)
                            .SetInput("data",A5)
                            .CreateSymbol("L5Bar");
                auto A5Bar = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L5Bar)
                            .CreateSymbol("A5Bar");

                auto L4Bar = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[3])
                            .SetParam("no_bias",false)
                            .SetInput("data",A5Bar)
                            .CreateSymbol("L4Bar");
                auto A4Bar = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L4Bar)
                            .CreateSymbol("A4Bar");

                auto L3Bar = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[2])
                            .SetParam("no_bias",false)
                            .SetInput("data",A4Bar)
                            .CreateSymbol("L3Bar");
                auto A3Bar = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L3Bar)
                            .CreateSymbol("A3Bar");

                auto L2Bar = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[1])
                            .SetParam("no_bias",false)
                            .SetInput("data",A3Bar)
                            .CreateSymbol("L2Bar");
                auto A2Bar = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L2Bar)
                            .CreateSymbol("A2Bar");

                auto L1Bar = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[0])
                            .SetParam("no_bias",false)
                            .SetInput("data",A2Bar)
                            .CreateSymbol("L1Bar");
                auto A1Bar = Operator("Activation")
                            .SetParam("act_type","relu")
                            .SetInput("data",L1Bar)
                            .CreateSymbol("A1Bar");

                auto Id = Operator("FullyConnected")
                            .SetParam("num_hidden",Sizes[0])
                            .SetParam("no_bias",false)
                            .SetInput("data",A1Bar)
                            .CreateSymbol("Id");
                //auto Sm = Operator("softmax")
                            //.CreateSymbol("Sm");

                auto Net = Operator("LinearRegressionOutput")
                            //.SetParam("ignore_label", -1)
                            .SetParam("grad_scale", 1)
                            //.SetParam("use_ignore", false)
                            //.SetParam("normalization", "null") /*batch,null,valid */
                            .SetInput("data", Id)
                            .SetInput("label",target_label)
                            .CreateSymbol("Net");

Thanks in advance, I can add any more information if required.

**Comment by aravindhv10** [comment]
Hello,

     I managed to figure out something, please inform me if this is right or wrong:
to get softmax of each data point (this is a 40 dimensional array) in a batch size of 256, I use softmax(axis=1) so this implemented as:

auto Sm = Operator("softmax")
                            .SetParam("axis", 1)
                            .SetInput("data", Id)
                            .CreateSymbol("Sm");

                auto Net = Operator("LinearRegressionOutput")
                            //.SetParam("ignore_label", -1)
                            .SetParam("grad_scale", 1)
                            //.SetParam("use_ignore", false)
                            //.SetParam("normalization", "null") /*batch,null,valid */
                            .SetInput("data", Sm)
                            .SetInput("label",target_label)
                            .CreateSymbol("Net");

**Comment by aravindhv10** [comment]
Hello again,

the above solution is not working as the network is not learning at all.

**Comment by frankfliu** [comment]
@mxnet-label-bot [Gluon, Question]
